Using the `dotnet msbuild` command produces invalid executables for apps targeting .NET Framework

After upgrading our TeamCity installation, I followed JetBrains recommendataion to [migrate from their Visual Studio (sln) runner﻿](https://www.jetbrains.com/help/teamcity/net.html#migrating-to-net-from-sln) to the new .NET runner. This basically means running `dotnet msbuild` instead of `MSBuild.exe`. Also, simply installing the .NET 5 SDK without having to deal with the complexity of [Visual Studio Build Tools](https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools) installation on the build agent seemed very appealing.

So I started to try building apps with `dotnet msbuild` instead of the usual `MSBuild.exe` command but had a very unpleasant surprise. The first app I tried was crashing at startup before even hitting the module initializer. The event viewer was showing a crash but without much usable details. When removing Costura, the app would stop crashing. So I [enabled fusion logs](https://stackoverflow.com/questions/255669/how-to-enable-assembly-bind-failure-logging-fusion-in-net/3256753#3256753) to try to understand what was happening and saw something very unexpected for a .NET Framework app: a reference to `System.Private.CoreLib`! (This is basically the .NET Core equivalent of mscorlib.)

Here is how to reproduce the problem with a minimal *hello world* app.

**Program.cs**:

```csharp
using System;

namespace hellocostura
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}
```

**hellocostura.csproj**:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net48</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Costura.Fody" Version="5.2.0" PrivateAssets="all" />
  </ItemGroup>

</Project>
```

**assemblyrefs.csx** (to check what are the referecend assemblies, runnable with the [dotnet script](https://github.com/filipw/dotnet-script) global tool):

```csharp
#r "nuget: dnlib, 3.3.2"

using (var stream = System.IO.File.OpenRead(Args[0]))
using (var module = dnlib.DotNet.ModuleDefMD.Load(stream))
{
    Console.WriteLine(string.Join("\n", module.GetAssemblyRefs()));
}
```

✅ When building with the standard msbuild exectuable, the `hellocostura.exe` has references to `mscorlib` and `System`, everything is fine.

```
rm -rf bin obj && \
  dotnet restore --nologo --verbosity=quiet && \
  msbuild -nologo -verbosity:quiet && \
  dotnet script assemblyrefs.csx bin/Debug/net48/hellocostura.exe
mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
```

❌ When building with the `dotnet msbuild` command, the `hellocostura.exe` has references to `mscorlib` and `System.Private.CoreLib`, leading to the crash at startup.

```
rm -rf bin obj && \
  dotnet restore --nologo --verbosity=quiet && \
  dotnet msbuild -nologo -verbosity:quiet && \
  dotnet script assemblyrefs.csx bin/Debug/net48/hellocostura.exe
mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
System.IO.Compression, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System.Runtime, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```

Note: the only difference bewtween the two commands is `msbuild` vs `dotnet msbuild`.

I will investigate this issue and hopefully be able to provide a workaround, maybe even another pull request. 😉

I don't think it has anything to do with the code in the templates. What I have noticed is that when using `dotnet msbuild` Cecil takes the *NET_CORE* path in [BaseAssemblyResolver.cs:137](https://github.com/jbevain/cecil/blob/0.11.3/Mono.Cecil/BaseAssemblyResolver.cs#L137-L139):

```csharp
#if NET_CORE
assembly = SearchTrustedPlatformAssemblies (name, parameters);
if (assembly != null)
    return assembly;
#else
```

When using `MSBuild.exe` it takes the other path (#else branch) and eventually [resolves the assembly in the GAC](https://github.com/jbevain/cecil/blob/0.11.3/Mono.Cecil/BaseAssemblyResolver.cs#L158).

So I tried to force Fody to use the `netclassictask` assembly like this:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net48</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Costura.Fody" Version="5.2.0" PrivateAssets="all" />
    <PackageReference Include="Fody" Version="6.5.1" PrivateAssets="all" GeneratePathProperty="true" />
  </ItemGroup>

  <PropertyGroup>
    <FodyAssembly>$(PkgFody)\netclassictask\Fody.dll</FodyAssembly>
  </PropertyGroup>

</Project>
```

But this did not work as it ended in an MSBuild error:

> %USERPROFILE%\\.nuget\packages\fody\6.5.1\build\Fody.targets(40,5): error MSB4062: The "Fody.WeavingTask" task could not be loaded from the assembly %USERPROFILE%\\.nuget\packages\fody\6.5.1\netclassictask\Fody.dll. **Could not load file or assembly 'Microsoft.Build.Utilities.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.** The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.

I think the solution may be to write a custom [IAssemblyResolver](https://github.com/jbevain/cecil/blob/0.11.3/Mono.Cecil/MetadataResolver.cs#L17:19).

--- a/src/Costura.Fody.Tests/Helpers/WeavingHelper.cs
+++ b/src/Costura.Fody.Tests/Helpers/WeavingHelper.cs
@@ -1,5 +1,4 @@
-﻿using System;
-using System.IO;
+﻿using System.IO;
 using System.Linq;
 using System.Xml.Linq;
 using Costura.Fody.Tests;
@@ -16,7 +15,6 @@ public static class WeavingHelper
             Config = XElement.Parse(config),
             References = string.Join(";", references.Select(r => Path.Combine(currentDirectory, r))),
             ReferenceCopyLocalPaths = references.Select(r => Path.Combine(currentDirectory, r)).ToList(),
-            AssemblyResolver = new TestAssemblyResolver()
         };
 
         if (!Path.IsPathRooted(assemblyPath))

--- a/src/Costura.Fody/AssemblyLoaderImporter.cs
+++ b/src/Costura.Fody/AssemblyLoaderImporter.cs
@@ -26,7 +26,7 @@ public partial class ModuleWeaver
     {
         var readerParameters = new ReaderParameters
         {
-            AssemblyResolver = AssemblyResolver,
+            AssemblyResolver = ModuleDefinition.AssemblyResolver,
             ReadSymbols = true,
             SymbolReaderProvider = new PdbReaderProvider(),
             SymbolStream = GetType().Assembly.GetManifestResourceStream("Costura.Template.pdb"),

--- a/src/Costura.Fody/ModuleWeaver.cs
+++ b/src/Costura.Fody/ModuleWeaver.cs
@@ -1,12 +1,9 @@
 ﻿using System.Collections.Generic;
 using System.Diagnostics;
 using Fody;
-using Mono.Cecil;
 
 public partial class ModuleWeaver: BaseModuleWeaver
 {
-    public IAssemblyResolver AssemblyResolver { get; set; }
-
     public override void Execute()
     {
 //#if DEBUG

